alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+
+        Parameters
+        ---------
+        X
+            Reference dataset to be summarized.
+        X_labels
+            Labels of the reference dataset.
+        Y
+            Dataset to choose the prototypes from. If ``None``, the prototypes will be selected from the reference
+            dataset `X`.
+
+        Returns
+        -------
+        self
+            Reference to itself.
+        """
+        self.X = X
+        # if the `X_labels` are not provided, then consider that all elements belong to the same class. This means
+        # that loss term which tries to avoid including in an epsilon ball elements belonging to other classes
+        # will always be 0. Still the first term of the loss tries to cover as many examples as possible with
+        # minimal overlap between the epsilon balls corresponding to the other prototypes.
+        self.X_labels = X_labels.astype(np.int32) if (X_labels is not None) else np.zeros((len(X), ), dtype=np.int32)
+        # if the set of prototypes is not provided, then find the prototypes belonging to the reference dataset.
+        self.Y = Y if (Y is not None) else self.X
+        # initialize penalty for adding a prototype
+        if self.lbd is None:
+            self.lbd = 1 / len(self.X)
+            self.meta['params'].update({'lbd': self.lbd})
+
+        self.max_label = np.max(self.X_labels)
+        self.kmatrix_yx = batch_compute_kernel_matrix(x=self.Y,
+                                                      y=self.X,
+                                                      kernel=self.kernel_distance,
+                                                      batch_size=self.batch_size,
+                                                      preprocess_fn=self.preprocess_fn)
+
+        return self
+
+    def explain(self, num_prototypes: int = 1) -> Explanation:
+        """
+        Searches for the requested number of prototypes. Note that the algorithm can return a lower number of
+        prototypes than the requested one. To increase the number of prototypes, reduce the epsilon-ball radius
+        `eps` and the penalty `lbd` for adding a prototype.
+
+        Parameters
+        ----------
+        num_prototypes
+            Number of maximum prototypes to be selected.
+
+        Returns
+        -------
+        An `Explanation` object containing the prototypes, prototypes indices and protoypes labels with additional \
+        metadata as attributes
+        """
+        if num_prototypes > len(self.Y):
+            num_prototypes = len(self.Y)
+            logger.warning('The number of prototypes requested is larger than the number of elements from '
+                           f'the prototypes selection set. Automatically setting `num_prototypes={num_prototypes}`.')
+
+        # dictionary of prototypes indices for each class
+        protos = {l: [] for l in range(self.max_label + 1)}  # type: Dict[int, List[int]]  # noqa: E741
```

Comment:
It seems that the implicit assumption we're making here is that labels are always represented as integers in `[0, ..., n-1]` and all labels are represented? Should we make this explicit in the description of `X_labels`?

Reply:
The assumption of integer representation is correct, but I think this is quite standard. 
On the other hand, we do not need all labels to be represented. It is true that there is some memory allocation that considers all the labels (I believe might be improved by adding an intermediate mapping and allocate the memory only for the one available), but we definitely do not assume that all labels from `0` to `n-1` are available. Might be the case that only `0, 1 & n-1` are there, which works just fine.

Reply:
check what happens if the labels do not follow `[0, L-1]` -> memory penalty and runtime